File: apps/web-dashboard/lib/api/transactions.ts
import { transactionClient } from "./client"
import type {
  AccountCreate,
  AccountResponse,
  CategoryCreate,
  CategoryResponse,
  PaginatedList,
  TransactionCreate,
  TransactionFilters,
  TransactionResponse,
  TransactionUpdate,
} from "@/types/api"

// ── Accounts ─────────────────────────────────────────────────────────────────

export async function getAccounts(): Promise<AccountResponse[]> {
  const res = await transactionClient.get<AccountResponse[]>("/accounts")
  return res.data
}

export async function createAccount(data: AccountCreate): Promise<AccountResponse> {
  const res = await transactionClient.post<AccountResponse>("/accounts", data)
  return res.data
}

export async function deleteAccount(id: string): Promise<void> {
  await transactionClient.delete(`/accounts/${id}`)
}

// ── Categories ────────────────────────────────────────────────────────────────

export async function getCategories(type?: "income" | "expense"): Promise<CategoryResponse[]> {
  const res = await transactionClient.get<CategoryResponse[]>("/categories", {
    params: type ? { type } : undefined,
  })
  return res.data
}

export async function createCategory(data: CategoryCreate): Promise<CategoryResponse> {
  const res = await transactionClient.post<CategoryResponse>("/categories", data)
  return res.data
}

// ── Transactions ──────────────────────────────────────────────────────────────

export async function getTransactions(
  filters: TransactionFilters = {}
): Promise<PaginatedList<TransactionResponse>> {
  const params: Record<string, unknown> = {}
  if (filters.account_id)  params.account_id  = filters.account_id
  if (filters.category_id) params.category_id = filters.category_id
  if (filters.type)        params.type        = filters.type
  if (filters.date_from)   params.date_from   = filters.date_from
  if (filters.date_to)     params.date_to     = filters.date_to
  params.skip  = filters.skip  ?? 0
  params.limit = filters.limit ?? 20

  const res = await transactionClient.get<PaginatedList<TransactionResponse>>(
    "/transactions",
    { params }
  )
  return res.data
}

export async function createTransaction(data: TransactionCreate): Promise<TransactionResponse> {
  const res = await transactionClient.post<TransactionResponse>("/transactions", data)
  return res.data
}

export async function updateTransaction(
  id: string,
  data: TransactionUpdate
): Promise<TransactionResponse> {
  const res = await transactionClient.put<TransactionResponse>(`/transactions/${id}`, data)
  return res.data
}

export async function deleteTransaction(id: string): Promise<void> {
  await transactionClient.delete(`/transactions/${id}`)
}

File: services/transaction-service/routers/transactions.py
import uuid
from datetime import datetime

from db import get_db
from fastapi import APIRouter, Depends, Query, status
from models.org import OrgMembership
from schemas.transaction import (
    TransactionCreate,
    TransactionListResponse,
    TransactionResponse,
    TransactionUpdate,
)
from services.security import get_org_member, get_write_member
from services.transaction_service import (
    create_transaction,
    delete_transaction,
    get_transaction,
    list_transactions,
    update_transaction,
)
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter(prefix="/transactions", tags=["transactions"])


@router.post("", response_model=TransactionResponse, status_code=status.HTTP_201_CREATED)
async def create(
    data: TransactionCreate,
    membership: OrgMembership = Depends(get_write_member),
    db: AsyncSession = Depends(get_db),
):
    return await create_transaction(membership.user_id, membership.org_id, data, db)


@router.get("", response_model=TransactionListResponse)
async def list_all(
    account_id: uuid.UUID | None = Query(None),
    category_id: uuid.UUID | None = Query(None),
    type: str | None = Query(None, description="income, expense, or transfer"),
    date_from: datetime | None = Query(None),
    date_to: datetime | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    membership: OrgMembership = Depends(get_org_member),
    db: AsyncSession = Depends(get_db),
):
    items, total = await list_transactions(
        membership.org_id,
        db,
        account_id=account_id,
        category_id=category_id,
        txn_type=type,
        date_from=date_from,
        date_to=date_to,
        skip=skip,
        limit=limit,
    )
    return TransactionListResponse(items=items, total=total, skip=skip, limit=limit)


@router.get("/{transaction_id}", response_model=TransactionResponse)
async def get_one(
    transaction_id: uuid.UUID,
    membership: OrgMembership = Depends(get_org_member),
    db: AsyncSession = Depends(get_db),
):
    return await get_transaction(transaction_id, membership.org_id, db)


@router.put("/{transaction_id}", response_model=TransactionResponse)
async def update(
    transaction_id: uuid.UUID,
    data: TransactionUpdate,
    membership: OrgMembership = Depends(get_write_member),
    db: AsyncSession = Depends(get_db),
):
    return await update_transaction(transaction_id, membership.org_id, data, db)


@router.delete("/{transaction_id}", status_code=status.HTTP_200_OK)
async def delete(
    transaction_id: uuid.UUID,
    membership: OrgMembership = Depends(get_write_member),
    db: AsyncSession = Depends(get_db),
):
    await delete_transaction(transaction_id, membership.org_id, db)
    return {"detail": "Transaction deleted"}
